move-vm-support/src/ss58_address.rs

```diff
@@ -0,0 +1,114 @@
+//! SS58 address format converter for Substrate and Move accounts.
+//! Based on the Pontem solution.
+
+use anyhow::{anyhow, ensure, Result};
+use blake2::{Blake2b512, Digest};
+use move_core_types::account_address::AccountAddress;
+
+// Substrate address prefix
+// Read more: https://docs.substrate.io/reference/address-formats/
+const SS58_PREFIX: &[u8] = b"SS58PRE";
+
+// Public key length in bytes
+const PUB_KEY_LENGTH: usize = 32;
+
+// Checksum length in bytes
+const CHECK_SUM_LEN: usize = 2;
+
+// Blake2b512 hash length in bytes
+const HASH_LEN: usize = 64;
+
+/// Convert ss58 address string to Move address structure
+/// In case if such conversion is not possible, return error.
+/// ```
+/// use move_vm_support::ss58_address::ss58_to_move_address;
+/// let substrate_address = "gkNW9pAcCHxZrnoVkhLkEQtsLsW5NWTC75cdAdxAMs9LNYCYg";
+/// let move_address = ss58_to_move_address(substrate_address).unwrap();
+/// assert_eq!(
+///    "0x8EAF04151687736326C9FEA17E25FC5287613693C912909CB226AA4794F26A48",
+///   format!("{:#X}", move_address)
+/// );
+/// ```
+pub fn ss58_to_move_address(ss58: &str) -> Result<AccountAddress> {
+    let bs58 = bs58::decode(ss58).into_vec()?;
+
+    ensure!(
+        bs58.len() > PUB_KEY_LENGTH + CHECK_SUM_LEN,
+        format!(
+            "Address length must be equal or greater than {} bytes",
+            PUB_KEY_LENGTH + CHECK_SUM_LEN
+        )
+    );
+
+    let check_sum = &bs58[bs58.len() - CHECK_SUM_LEN..];
+    let address = &bs58[bs58.len() - PUB_KEY_LENGTH - CHECK_SUM_LEN..bs58.len() - CHECK_SUM_LEN];
+
+    if check_sum != &ss58_hash(&bs58[0..bs58.len() - CHECK_SUM_LEN])[0..CHECK_SUM_LEN] {
+        return Err(anyhow!("Wrong address checksum"));
+    }
+
+    let mut addr = [0; PUB_KEY_LENGTH];
+    addr.copy_from_slice(address);
+    Ok(AccountAddress::new(addr))
+}
+
+/// Convert SS58 address to Move address string.
+pub fn ss58_to_move_address_string(ss58: &str) -> Result<String> {
+    Ok(format!("{:#X}", ss58_to_move_address(ss58)?))
+}
+
+// Helper function which calculates the BLAKE2b512 hash of the given data.
+fn ss58_hash(data: &[u8]) -> [u8; HASH_LEN] {
```

Comment:
We could rename it to `ss58_checksum` and then return directly `[u8; CHECKSUM_LEN]` to simplify our use case above.

Reply:
Yes, we could. And I even thought about it for a while.

But.

In case we'd like to make the opposite conversion with a function like:
```
pub fn move_address_to_ss58(addr: &AccountAddress) -> String {
    let mut ss58_address = [0; 35];
    ss58_address[0] = 42;
    ss58_address[1..33].copy_from_slice(&addr.to_vec());
    let hash = ss58_hash(&ss58_address[0..33]);
    ss58_address[33..35].copy_from_slice(&hash[0..2]);
    bs58::encode(ss58_address).into_string()
}
```
we'd need hash, not checksum only.

I'm still curious if there is a need for an opposite conversion inside the compiler.

Reply:
> I'm still curious if there is a need for an opposite conversion inside the compiler.

Hopefull we are gonna need that :smile:

Reply:
Yes, I've just got fixed on returning whole hash.

I'll change it in a minute. If there will be something more needed we can always extend it and introduce new function.

Reply:
>     I'm still curious if there is a need for an opposite conversion inside the compiler.
> 
> Hopefully, we are not gonna need that 😄
> 

For now, the only thing I can imaging is when something is gonna happen inside, eg. Prover, which operates on translated addresses, and we'd like to present something to the user. Then probably we'd need that conversion because users operate on Substrate addresses.
But that's the future 😄

Reply:
Changed.